Adopt more smart pointers in LocalDOMWindow

rwlbuis opened the pull request:
#### b838f8bb3573bd5906bc5f02fcc8cb274b3c9b8a
<pre>
Adopt more smart pointers in LocalDOMWindow
<a href="https://bugs.webkit.org/show_bug.cgi?id=273393">https://bugs.webkit.org/show_bug.cgi?id=273393</a>

Reviewed by Chris Dumez.

Adopt more smart pointers in LocalDOMWindow based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::allowsBeforeUnloadListeners):
(WebCore::LocalDOMWindow::allowPopUp):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::LocalDOMWindow::collectMatchingElementsInFlatTree):
(WebCore::LocalDOMWindow::matchingElementInFlatTree):
(WebCore::LocalDOMWindow::closePage):
(WebCore::LocalDOMWindow::origin const):
(WebCore::LocalDOMWindow::securityOrigin const):

Canonical link: <a href="https://commits.webkit.org/281244@main">https://commits.webkit.org/281244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a9b35e44333bbd41b9c7a9c1b31b09b60784c42

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59249 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63163 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9692 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46246 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9922 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6913 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61279 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36066 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51287 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28960 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32779 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8545 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8696 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54737 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64862 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3177 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3188 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51287 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/55558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13125 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2652 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34408 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35236 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -640,9 +640,14 @@ ExceptionOr<Ref<NodeList>> LocalDOMWindow::collectMatchingElementsInFlatTree(Nod
-    for (auto& node : composedTreeDescendants(*scopeContainer)) {
-        if (RefPtr element = dynamicDowncast<Element>(node); element && query.matches(*element) && !node.isInUserAgentShadowTree())
+    auto descendants = composedTreeDescendants(*scopeContainer);
+    auto it = descendants.begin();
```
Unclear why the new code looks so much worse here.

rwlbuis replied:
I had some difficulty reffing  the node and at the same time appending correctly to result, hence the complicated code. But I think we can keep it simpler and still fix the warning, by simply creating a StaticNodeList instead of a StaticElementList, since in the end we return a NodeList.

cdumez replied:
I doubt these changes are needed, can you please explain what compiler error you were getting?

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2763,7 +2768,7 @@ void LocalDOMWindow::disableSuddenTermination()
 LocalFrame* LocalDOMWindow::frame() const
 {
-    auto* document = this->document();
+    SUPPRESS_UNCOUNTED_LOCAL auto* document = this->document();
```
Please do not add SUPPRESS_UNCOUNTED_LOCAL.

rwlbuis replied:
Removed.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -639,13 +639,13 @@ ExceptionOr<Ref<NodeList>> LocalDOMWindow::collectMatchingElementsInFlatTree(Nod
-            result.append(element.releaseNonNull());
+    Vector<Ref<Node>> result;
+    for (Ref node : composedTreeDescendants(*scopeContainer)) {
+        if (auto* element = dynamicDowncast<Element>(node.get()); element && query.matches(*element) && !node->isInUserAgentShadowTree())
```
Can we use the following here?
```
if (RefPtr element = dynamicDowncast<Element>(WTFMove(node)); element && query.matches(*element) && !element->isInUserAgentShadowTree())
    result.append(WTFMove(element));
```

And then keep Element type in the Vector? I don't understand why we're switching to Node type here.

rwlbuis replied:
Sorry for the delay. It turned out the warning can be fixed in a simpler way.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -955,7 +955,7 @@ void LocalDOMWindow::processPostMessage(JSC::JSGlobalObject& lexicalGlobalObject
             return;
 
-        auto& vm = globalObject->vm();
+        Ref vm = globalObject->vm();
```
I think the JSC team has expressed concerns about ref'ing the VM unnecessarily?

rwlbuis replied:
Got it, I removed this part.

rwlbuis committed:
Adopt more smart pointers in LocalDOMWindow
https://bugs.webkit.org/show_bug.cgi?id=273393

Reviewed by Chris Dumez.

Adopt more smart pointers in LocalDOMWindow based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::allowsBeforeUnloadListeners):
(WebCore::LocalDOMWindow::allowPopUp):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::LocalDOMWindow::collectMatchingElementsInFlatTree):
(WebCore::LocalDOMWindow::matchingElementInFlatTree):
(WebCore::LocalDOMWindow::closePage):
(WebCore::LocalDOMWindow::origin const):
(WebCore::LocalDOMWindow::securityOrigin const):

Canonical link: https://commits.webkit.org/281244@main
